skbio/stats/composition.py

```diff
@@ -0,0 +1,359 @@
+
+This module allows two styles of manipulation of compositional data.
+Compositional data can be analyzed using perturbation and power
+operations, which can be useful for simulation studies. The
+alternative strategy is to transform compositional data into the real
+space.  Right now, the centre log ratio transform (clr) [1]_ can be
+used to accomplish this.  This transform can be useful for performing
+standard statistical tools such as parametric hypothesis testing,
+regressions and more.
+
+The major caveat of using this framework is dealing with zeros.  In
+the Aitchison geometry, only compositions with nonzero components can
+be considered. The multiplicative replacement technique [2]_ can be
+used to substitute these zeros with small pseudocounts without
+introducing major distortions to the data.
+
+Functions
+---------
+
+.. autosummary::
+   :toctree: generated/
+
+   closure
+   multiplicative_replacement
+   perturb
+   perturb_inv
+   power
+   clr
+   centralize
+
+Examples
+--------
+
+>>> import numpy as np
+
+Consider a very simple environment with only 3 species. The species
+in the environment are equally distributed and their proportions are
+equivalent:
+
+>>> otus = np.array([1./3, 1./3., 1./3])
+
+Suppose that an antibiotic kills off half of the population for the
+first two species, but doesn't harm the third species. Then the
+perturbation vector would be as follows
+
+>>> antibiotic = np.array([1./2, 1./2, 1])
+
+And the resulting perturbation would be
+
+>>> perturb(otus, antibiotic)
+array([ 0.25,  0.25,  0.5 ])
+
+Reference
+---------
+.. [1] V. Pawlowsky-Glahn. "Lecture Notes on Compositional Data Analysis"
+.. [2] J. A. Martin-Fernandez. "Dealing With Zeros and Missing Values in
+       Compositional Data Sets Using Nonparametric Imputation"
+
+"""
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+import numpy as np
+import scipy.stats as ss
+
+
+def closure(mat):
+    """
+    Performs closure to ensure that all elements add up to 1.
+
+    Parameters
+    ----------
+    mat : array_like
+       a matrix of proportions where
+       rows = compositions
+       columns = components
+
+    Returns
+    -------
+    array_like, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    """
+    mat = np.atleast_2d(mat)
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    mat = mat.astype(np.float64)
```

Comment:
It'd be simpler to just call `mat.astype(np.float64, copy=False)` instead of using an if statement. That way we can let numpy determine whether to make a copy or not.